src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs

```diff
@@ -153,8 +154,19 @@ public void Stop()
                             }
                         }
 
-                        await requestHandler(typedParams, requestContext);
-                        Logger.Verbose($"Finished processing message with id[{requestMessage.Id}], of type[{requestMessage.MessageType}] and method[{requestMessage.Method}]");
+                        if (isBackgroundTask)
+                        {
+                            // Start a detached task that runs in background, unblocking dispatcher thread.
+                            await Task.Factory.StartNew(() => Task.Run(async () =>
```

Comment:
Why bother nesting the calls here vs just calling Task.Run and not awaiting? These should be doing the same thing so it seems unnecessary.

https://devblogs.microsoft.com/pfxteam/task-run-vs-task-factory-startnew/

Reply:
Not really, if you run a simple test, you'll notice Task.Run doesn't detach the thread from the main thread, whereas Task.Factory.StartNew does.

Link to example: https://gist.github.com/cheenamalhotra/b47770f5fbc7330aa79968a8999c8884

Reply:
You have a bug in your example, `Task.Factory.StartNew` returns a `Task<TResult>`. In your case the TResult is a task itself, so the return type is `Task<Task>`, but since you only await once it never awaits the inner task (the delay).

If you add a second await (or call `.Unwrap()` then both of them behave the same : 

`await Run("Task.Factory.StartNew", async () => await await Task.Factory.StartNew(async () => await PrintAndDelayAsync("Task.Factory.StartNew", 1)));`

This blob explains it better : https://code-maze.com/csharp-task-run-vs-task-factory-startnew/ (the `The async Awareness
` section)

> On top of this, StartNew wraps this in a Task construct. Eventually, we end up with a Task<Task<string>> instance which is not what we desire.



So what you have is essentially the same as calling Task.Run directly and not awaiting the task it returns.

Reply:
And actually that blog clearly states that Task.Run should be preferred : 

> Now, let’s talk about the difference from the TaskScheduler context. Task.Run(action) internally uses the default TaskScheduler, which means it always offloads a task to the thread pool.  StartNew(action), on the other hand, uses the scheduler of the current thread which may not use thread pool at all! 

> This can be a matter of concern particularly when we work with the UI thread! If we initiate a task by StartNew(action) within a UI thread, it will utilize the scheduler of the UI thread and no offloading happens. That means, if the task is a long-running one, UI will soon become irresponsive. Task.Run is free from this risk as it will always offload work to the thread pool no matter in which thread it has been initiated. So, Task.Run is the safer option in such cases.

Reply:
> And actually that blog clearly states that Task.Run should be preferred :

I wasn't aware of the thread scheduler defaults being a concern, so thanks for sharing that. In that case we can explicitly specify the `TaskScheduler` when calling `StartNew` to address that.

Using `Task.Run` is an alternative but we'll have to disable/re-enable C# Pragma warning to not await on it to make it truly non-blocking which I'm trying to avoid. I'll update the defaults in `Task.Factory.StartNew` to ensure we're doing the right thing.

Reply:
You can also assign it to a discard var 

e.g. 

`_ = Task.Run(async () =>` 

https://github.com/microsoft/sqltoolsservice/blob/16c7e323567023b08200ab35b2bfb2a5042385e0/src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs#L353

Reply:
I have a very slight preference to just always use Task.Run since it's simpler and will help avoid someone copying it and forgetting to add the correct set of defaults. But not a blocking concern if you'd prefer to use StartNew